src/core/common/trickle_timer.hpp

```diff
@@ -102,60 +102,85 @@ class TrickleTimer : public TimerMilli
+    void Start(Mode     aMode,
+               uint32_t aIntervalMin,
+               uint32_t aIntervalMax,
+               uint16_t aRedundancyConstant = kInfiniteRedundancyConstant);
 
     /**
      * This method stops the trickle timer.
      *
      */
-    void Stop(void);
+    void Stop(void) { TimerMilli::Stop(); }
 
-#ifdef ENABLE_TRICKLE_TIMER_SUPPRESSION_SUPPORT
     /**
-     * This method indicates to the trickle timer a 'consistent' state.
+     * This method indicates to the trickle timer a 'consistent' event.
+     *
+     * The 'consistent' events are used to control suppression behavior. The trickle timer keeps track of number of
+     * 'consistent' event in each interval. The timer handler is invoked only if the number of `consistent` events
+     * received in the interval is less than the redundancy constant.
      *
      */
-    void IndicateConsistent(void) { mCounter++; }
-#endif
+    void IndicateConsistent(void);
 
     /**
-     * This method indicates to the trickle timer an 'inconsistent' state.
+     * This method indicates to the trickle timer an 'inconsistent' event.
+     *
+     * Receiving an 'inconsistent' event causes the trickle timer to reset (i.e., start with interval set to the min
+     * value) unless the current interval being used is already equal to the min interval.
      *
      */
     void IndicateInconsistent(void);
 
 private:
+    enum Phase : uint8_t
+    {
+        kBeforeRandomTime, // Trickle timer is before random time `t` in the current interval.
+        kAfterRandomTime,  // Trickle timer is after random time `t` in the current interval.
+    };
+
     void        StartNewInterval(void);
     static void HandleTimer(Timer &aTimer);
     void        HandleTimer(void);
     void        HandleEndOfTimeInInterval(void);
     void        HandleEndOfInterval(void);
-    void        StartAt(void) {} // Shadow base class `TimerMilli` method to ensure it is hidden.
-
-#ifdef ENABLE_TRICKLE_TIMER_SUPPRESSION_SUPPORT
-    const uint32_t mRedundancyConstant; // Redundancy constant (aka 'k').
-    uint32_t       mCounter;            // A counter for number of "consistent" transmissions (aka 'c').
-#endif
-
-    uint32_t mIntervalMin;            // Minimum interval (aka `Imin`).
-    uint32_t mIntervalMax;            // Maximum interval (aka `Imax`).
-    uint32_t mInterval;               // Current interval (aka `I`).
-    uint32_t mTimeInInterval;         // Time in interval (aka `t`).
-    Handler  mTransmitHandler;        // Transmit handler callback.
-    Handler  mIntervalExpiredHandler; // Interval expired handler callback.
-    Mode     mMode;                   // Trickle timer mode.
-    bool     mIsRunning : 1;          // Indicates if the trickle timer is running.
-    bool     mInTransmitPhase : 1;    // Indicates if in transmit phase (before time `t` in current interval `I`).
+
+    // Shadow base class `TimerMilli` methods to ensure they are hidden.
```

Comment:
Can we make `TimerMilli` private?

Reply:
I was also thinking of the same and wanted to see if we can improve this 😃 
But I could not find a good way for it:

-  First approach I considered was instead of inheriting from `TimerMilli` we define it as `mTimer` member variable
   - The problem with this is that from the callback of `mTimer`, we cannot back to the owning the `TrickleTimer` object (we can use `offsetof()` with pointer tricks but not safe/recommended for all tool-chains).

- We cannot define the inheritance from `TimerMilli` as `private` because we need to publicly inherit the `InstanceLocator` methods (which are used in the callback to get back to the object owning the `TrickleTimer`).
   - We can try to use `virtual` inheritance to allow multiple base class inheritance from `InstacneLocator` but that itself adds its own overheads.

- Another option I considered was to use a `mTimer` member variable but define it as `TimerMilliContext` (which explicily saves some context).
   - This works fine but increases the RAM size of `TrickleTimer`.

- Finally I opted to continue to use the existing model (inherit publicly form `TimerMill` but try to hide the methods that we don't want by shadow private definitions).